src/lightning/pytorch/loops/fetchers.py

```diff
@@ -43,16 +43,17 @@ def combined_loader(self) -> CombinedLoader:
 
     def setup(self, combined_loader: CombinedLoader) -> None:
         self._combined_loader = combined_loader
-        self.length = sized_len(combined_loader)
-        self.done = self.length == 0
 
     def __iter__(self) -> "_DataFetcher":
-        self.reset()
         self.iterator = iter(self.combined_loader)
+        self.reset()
         return self
 
     def __next__(self) -> Any:
         assert (iterator := self.iterator) is not None
+        if self.length is not None and self.fetched == self.length:
+            # The iterator may still have items, but the length is reached (determined by the limits)
+            raise StopIteration
```

Comment:
Why did you have to do this? I don't think this is right under the prefetching>0 case.

For instance, if we have 10 samples, we are processing the 5th one, and `prefetch_batches=5`, then `self.fetched` will be equal to `self.length`, but there are still 5 batches to return

If you didn't have this code, the except below should still trigger, and be caught by the `__next__` logic of `_PrefetchDataFetcher`.

If you could tell me which test fails without this addition, I could take a look to understand what's going on

Reply:
There are no old tests that fail without this addition. This was an additional restriction to ensure the set limits are respected. 

The additional assertion I added for this is in
`test_done_dataloader_iter_with_limit` at the bottom

```py
    if fetcher.done:
        with pytest.raises(StopIteration):
            next(dataloader_iter)
```

If you don't want me to do this in this PR fine, it can be done later. IMO now that limits are taken into consideration by the loop, I think it is fair to also enforce it for the user on the dataloader_iter.

Reply:
Okay. I agree with the objective, but it should be done by adding

`if fetcher.done: raise StopIteration` here: https://github.com/Lightning-AI/lightning/blob/master/src/lightning/pytorch/loops/fetchers.py#L191

Since this is specific to `dataloader_iter`